### ISSUE TITLE
Blocking "ran out of predicates" with maximum_comparisons = 500,000,000

### ISSUE THREAD

**Issue by wleftwich** [datamade/dedupe#473]
Dedupe version 1.4.14

With 1.7 million name&address records to process, I create an empty Dedupe object, load 100k sample records from the full dataset  with the Dedupe.sample() method, and run a training session till I get  30 positive and 30 negative examples. At the end of train() I get the "ran out of predicates" warning. If I go ahead and call do_matching(), recall is somewhere south of 10 percent.

EDIT 6/23 - s/do_matching()/StaticDedupe.matchBlocks()/

This happens even with maximum_comparisons = 500,000,000 and recall = 0.9.

However, when I provide a 100k-item data sample to the Dedupe constructor, so that the sample() method never gets called, training proceeds without the warning and do_matching() produces 10 times as many matches.

**Comment by fgregg** [comment]
* How are you taking your initial sample?
* What is `do_matching`?

Block learning uses the positive training examples and a random sample of 900 records. When you use the `sample` method, the records are sampled from the entire set of records. When you supply a sample, then the records are sampled from that sample. 

If the your initial sampling is random sample, then there should not be any expected difference between this one and two stage process.

**Comment by wleftwich** [comment]
- How are you taking your initial sample?

I'm using Dedupe.sample() to sample the entire data set, dumping a pickle of Dedupe.data_sample, and exiting. Then in another script I pickle.load() the sample and pass it to the Dedupe() constructor. I started doing this when I was testing various params and wanted the same sample every time.

- What is do_matching?

Sorry, that is my own not-very-well-named function. I meant StaticDedupe.matchBlocks().

- There should not be any expected difference between this one and two stage process.

OK, this is somewhat embarrassing because it involves code I submitted back in April and then forgot about (https://github.com/datamade/dedupe/pull/459).

The difference is in Dedupe.sampled_records.original_length. When you use Dedupe.sample(data), it's the length of the original data. When you use Dedupe(fields, data_sample) it's the length of the sample.

So at this point it looks like I am getting usable results when I do it The Wrong Way but not The Right Way.

**Comment by wleftwich** [comment]
I see now that DedupeBlockLearner.esitmate() increases as the square of Dedupe.sampled_records.original_length, so I'll try cranking maximum_comparisons still higher. As a workaround on getting a correct value into original_length when passing data_sample to Dedupe.__init__(), I'll just put the value into configuration and set it directly on the Dedupe object.

**Comment by wleftwich** [comment]
Just FYI,  I got usable results on a dataset of 7.7 million business contacts, running on an EC2 r3.xlarge box, 30 GB memory). recall=0.9, maximum_comparisons=9e9,  index_predicates=False. Blocking takes about 30 minutes, matching 20.

Thanks again for putting this library out there.

**Comment by itcraftsmanpl** [comment]
@wleftwich how many fields do you used, and how big was your sample ? I have 2.5 milion entities, 7 field, 40GB of ram and after several hours of running dedupe  it's killed by system (out of resources).

**Comment by wleftwich** [comment]
Here's the relevant section of my config file.
```
sample_size: 100000  # default 15000
maximum_comparisons: 9000000000  # default 1_000_000
recall:  0.95 # default 0.975
index_predicates: False # default True

fields:
  - {field: firstname, type: String}
  - {field: lastname, type: String}
  - {field: company, has_missing: true, type: String}
  - {field: address, type: String}
  - {field: city, type: ShortString}
  - {field: state, type: Exact}
  - {field: zip, type: ShortString}
  - {field: phones, has missing: true, type: String}
  - {field: emails, has missing: true, type: String}
```
I wrote two scripts, train.py and run_dedupe.py. 
For training, I pre-sampled my 7.7 million record dataset down to 770,000; then used the Dedupe.sample method to make a training sample of 100,000 records. This meant that after sampling I had to set Dedupe.sampled_records.original_length to 7.7 million. Otherwise the blocks become too large, which is what I suspect is happening to you.

During development I found it useful to log memory usage like this:
`logger.info("Max mem so far %s" % resource.getrusage(resource.RUSAGE_SELF).ru_maxrss)`

**Comment by fgregg** [comment]
I think this was largely due to the way that we did sampling which has been addressed in #477

**Comment by wleftwich** [comment]
Agreed, no problems since #477. Thanks!
